File: script_manager.py
#!/usr/bin/env python3
"""
Script Manager - Handles the pre-generated script database
Provides scripts to the video pipeline without API calls
"""

import json
import random
from datetime import datetime
from pathlib import Path

SCRIPTS_FILE = Path("data/bible_scripts.json")

class ScriptManager:
    def __init__(self):
        self.scripts_file = SCRIPTS_FILE
        self.data = self.load_scripts()
    
    def load_scripts(self):
        """Load scripts from database file"""
        if self.scripts_file.exists():
            try:
                with open(self.scripts_file, 'r', encoding='utf-8') as f:
                    return json.load(f)
            except Exception as e:
                print(f"Error loading scripts: {e}")
                return {"scripts": [], "generated_at": datetime.now().isoformat()}
        return {"scripts": [], "generated_at": datetime.now().isoformat()}
    
    def save_scripts(self):
        """Save scripts back to database"""
        with open(self.scripts_file, 'w', encoding='utf-8') as f:
            json.dump(self.data, f, indent=2, ensure_ascii=False)
    
    def get_next_script(self, mark_used=True):
        """Get the next unused script"""
        unused_scripts = [s for s in self.data["scripts"] if not s.get("used", False)]
        
        if not unused_scripts:
            return None
        
        # Get random unused script for variety
        script = random.choice(unused_scripts)
        
        if mark_used:
            # Mark as used
            for i, s in enumerate(self.data["scripts"]):
                if s["id"] == script["id"]:
                    self.data["scripts"][i]["used"] = True
                    self.data["scripts"][i]["used_at"] = datetime.now().isoformat()
                    break
            
            self.save_scripts()
        
        return script
    
    def get_script_stats(self):
        """Get statistics about script usage"""
        total = len(self.data["scripts"])
        used = len([s for s in self.data["scripts"] if s.get("used", False)])
        unused = total - used
        
        return {
            "total": total,
            "used": used,
            "unused": unused,
            "usage_percentage": (used / total * 100) if total > 0 else 0
        }
    
    def reset_usage(self):
        """Reset all scripts to unused (for testing)"""
        for script in self.data["scripts"]:
            script["used"] = False
            if "used_at" in script:
                del script["used_at"]
        
        self.save_scripts()
        print("✅ All scripts reset to unused")
    
    def preview_next_scripts(self, count=5):
        """Preview upcoming unused scripts"""
        unused_scripts = [s for s in self.data["scripts"] if not s.get("used", False)]
        
        if not unused_scripts:
            print("❌ No unused scripts available!")
            return
        
        print(f"📋 Next {min(count, len(unused_scripts))} unused scripts:")
        print("=" * 60)
        
        for i, script in enumerate(unused_scripts[:count]):
            print(f"\n🎬 Script {i+1} (ID: {script['id']}):")
            print(f"Style: {script.get('style', 'Unknown')}")
            print(f"Verse: {script.get('verse', 'Unknown')[:50]}...")
            print("Script Preview:")
            print(script['script'][:200] + ("..." if len(script['script']) > 200 else ""))
            print("-" * 40)

def get_script_for_pipeline():
    """
    Simple function for pipeline integration
    Returns just the script text, ready to use
    """
    manager = ScriptManager()
    script_data = manager.get_next_script()
    
    if not script_data:
        # Fallback script if database is empty
        return """HOOK: This Bible verse will inspire you today!
VERSE: Trust in the Lord with all your heart.
MEANING: Complete faith brings peace and direction.
APPLICATION: Choose trust over worry in every situation.
CTA: Share your testimony! #Bible #Faith #Trust #Inspiration"""
    
    return script_data["script"]

def check_script_supply():
    """Check if we need to generate more scripts"""
    manager = ScriptManager()
    stats = manager.get_script_stats()
    
    print(f"📊 Script Supply Status:")
    print(f"   Total scripts: {stats['total']}")
    print(f"   Used: {stats['used']}")
    print(f"   Remaining: {stats['unused']}")
    print(f"   Usage: {stats['usage_percentage']:.1f}%")
    
    if stats['unused'] < 10:
        print("\n⚠️  LOW SCRIPT SUPPLY!")
        print("   Run batch_script_generator.py to create more scripts")
        return False
    elif stats['unused'] < 25:
        print("\n🔔 Consider generating more scripts soon")
        return True
    else:
        print("\n✅ Good script supply")
        return True

if __name__ == "__main__":
    print("🎬 Bible Script Manager")
    print("=" * 30)
    
    manager = ScriptManager()
    
    # Show status
    check_script_supply()
    
    print("\nOptions:")
    print("1. Get next script")
    print("2. Preview upcoming scripts") 
    print("3. Show detailed stats")
    print("4. Reset all usage (testing)")
    print("5. Exit")
    
    while True:
        choice = input("\nEnter choice (1-5): ").strip()
        
        if choice == "1":
            script = manager.get_next_script()
            if script:
                print(f"\n🎬 Next Script (ID: {script['id']}):")
                print(f"Style: {script.get('style', 'Unknown')}")
                print(f"Verse: {script.get('verse', 'Unknown')}")
                print("\nScript:")
                print(script['script'])
                print("\n✅ Script marked as used")
            else:
                print("❌ No unused scripts available!")
                
        elif choice == "2":
            count = input("How many to preview? (default 5): ").strip()
            count = int(count) if count.isdigit() else 5
            manager.preview_next_scripts(count)
            
        elif choice == "3":
            stats = manager.get_script_stats()
            print(f"\n📊 Detailed Stats:")
            print(f"   Total scripts in database: {stats['total']}")
            print(f"   Scripts used: {stats['used']}")
            print(f"   Scripts remaining: {stats['unused']}")
            print(f"   Usage percentage: {stats['usage_percentage']:.1f}%")
            
        elif choice == "4":
            confirm = input("Reset all scripts to unused? (y/N): ").strip()
            if confirm.lower().startswith('y'):
                manager.reset_usage()
            
        elif choice == "5":
            break
            
        else:
            print("Invalid choice. Try again.")

File: script_manager.mjs
// Script Management Utility
// Use this to monitor, manage, and generate new Bible scripts

import { ScriptManager } from './pipeline_integration.mjs';
import fs from 'fs-extra';
import path from 'path';

const scriptManager = new ScriptManager();

async function showStatus() {
  console.log('📊 BIBLE SCRIPTS STATUS REPORT');
  console.log('='.repeat(50));
  
  const report = await scriptManager.getDetailedReport();
  const stats = report.stats;
  
  console.log(`📦 Total Scripts: ${stats.total}`);
  console.log(`✅ Used Scripts: ${stats.used}`);
  console.log(`📝 Unused Scripts: ${stats.unused}`);
  console.log(`📈 Usage Percentage: ${stats.usage_percentage.toFixed(1)}%`);
  console.log(`🔄 Average Uses per Script: ${stats.average_uses_per_script.toFixed(1)}`);
  console.log(`♻️  Scripts Used Multiple Times: ${stats.scripts_used_multiple_times}`);
  
  // Supply status
  const supplyCheck = await scriptManager.checkScriptSupply();
  if (supplyCheck.critical) {
    console.log('🚨 STATUS: CRITICAL - Generate scripts immediately!');
  } else if (supplyCheck.needsRefresh) {
    console.log('⚠️  STATUS: LOW - Consider generating more scripts');
  } else {
    console.log('✅ STATUS: GOOD - Plenty of scripts available');
  }
  
  console.log('\n🏆 MOST USED SCRIPTS:');
  report.mostUsed.forEach((script, i) => {
    console.log(`${i+1}. ${script.verse.substring(0, 40)}... (used ${script.usage_count} times)`);
  });
  
  console.log('\n🕒 RECENTLY USED SCRIPTS:');
  report.recentlyUsed.forEach((script, i) => {
    const date = new Date(script.used_at).toLocaleString();
    console.log(`${i+1}. ${script.verse.substring(0, 40)}... (${date})`);
  });
}

async function cleanupOldScripts() {
  console.log('🧹 CLEANUP: Finding scripts used multiple times...');
  
  const data = await scriptManager.loadScripts();
  const multipleUseScripts = data.scripts.filter(s => s.usage_count > 1);
  
  if (multipleUseScripts.length === 0) {
    console.log('✅ No scripts used multiple times found');
    return;
  }
  
  console.log(`Found ${multipleUseScripts.length} scripts used multiple times:`);
  multipleUseScripts.forEach(script => {
    console.log(`- ${script.verse.substring(0, 50)}... (used ${script.usage_count} times)`);
  });
  
  console.log('💡 TIP: You can manually review and delete these from data/bible_scripts.json');
}

async function resetAllUsage() {
  console.log('🔄 RESET: Marking all scripts as unused...');
  
  const data = await scriptManager.loadScripts();
  data.scripts.forEach(script => {
    script.used = false;
    delete script.used_at;
    delete script.usage_count;
  });
  
  await scriptManager.saveScripts(data);
  console.log(`✅ Reset complete! All ${data.scripts.length} scripts are now available for use.`);
}

async function generateNewScripts() {
  console.log('🤖 GENERATE: Starting new script generation...');
  console.log('💡 TIP: This would typically call your batch_script_generator.py');
  console.log('📝 For now, run: python batch_script_generator.py');
  console.log('🔄 Then the new scripts will be automatically merged into your existing collection');
}

// CLI Interface
async function handleCLI() {
  const command = process.argv[2];
  
  switch (command) {
    case 'status':
      await showStatus();
      break;
    case 'cleanup':
      await cleanupOldScripts();
      break;
    case 'reset':
      await resetAllUsage();
      break;
    case 'generate':
      await generateNewScripts();
      break;
    default:
      console.log('📋 BIBLE SCRIPTS MANAGER');
      console.log('Usage: node script_manager.mjs [command]');
      console.log('');
      console.log('Commands:');
      console.log('  status    - Show detailed script usage report');
      console.log('  cleanup   - Find scripts used multiple times');
      console.log('  reset     - Mark all scripts as unused (fresh start)');
      console.log('  generate  - Generate new scripts (calls Python generator)');
      console.log('');
      console.log('Examples:');
      console.log('  node script_manager.mjs status');
      console.log('  node script_manager.mjs cleanup');
      await showStatus(); // Show status by default
  }
}

// Run CLI
if (import.meta.url.endsWith(process.argv[1]) || process.argv[1].endsWith('script_manager.mjs')) {
  handleCLI().catch(console.error);
}
